containers/src/Data/IntMap/Internal.hs

```diff
@@ -13,6 +13,7 @@
 #endif
 
 {-# OPTIONS_HADDOCK not-home #-}
+{-# OPTIONS_GHC -fno-warn-incomplete-uni-patterns #-}
```

Comment:
If these warnings matter, why are they simply disabled?

Reply:
I'm not claiming they matter.

containers is a boot library. If it doesn't build then GHC will not build. These warnings [will be enabled by `-Wall` in futures GHCs](https://gitlab.haskell.org/ghc/ghc/-/issues/15656) therefore if containers triggers these warnings it will prevent GHC from building. I added these warnings to `ghc-options` not because they matter but because code triggering them shouldn't be added accidentally to containers.

Reply:
Put another way, it is a means of achieving forward compatibility.

Reply:
> I added these warnings to `ghc-options` not because they matter but because code triggering them shouldn't be added accidentally to containers.

How can we tell that potentially problematic code that would trigger the warnings is added, when the warning is disabled in several modules?

Reply:
Basically, disabled warnings are code smell to me. I think it would be nice if the proponents of the new warning would actually address the problems in the code.

Reply:
It's a very valid concern whether containers should contain code that is not clean for these warnings. However, that is not my concern in this particular PR.

My concern is that code in boot libraries (of which containers is one) should not trigger `incomplete-uni-patterns` and `incomplete-record-updates` *if thery were to be enabled*. That is because the plan is to enable these warnings in `-Wall`. As containers is built with `-Wall`, and GHC enables `-Werror` on its boot libraries, containers not being clean for these warnings will prevent them being added to `-Wall`.

To ensure that containers is clean for these warnings we have a few options, from most difficult to most simple

1. Explicitly enable the warnings in containers's cabal file, find all code that triggers them and rewrite it
2. Explicitly enable the warnings in container's cabal file, find all code that triggers them and disable the warnings on a per-module basis
3. Explicitly *dis*able the warnings in container's cabal file and do nothing to existing code

This PR implements option 2 because it has a nice combination of easy to implement and providing actual value to containers. If you prefer 3 then please let me know. I can do that. 1 is probably more controversial since it would require actually changing some of the code of containers.

But perhaps I'm not actually explaning sufficiently well the background to this change and why it is necessary. Would you like me to go into more detail?

Reply:
> it would be nice if the proponents of the new warning would actually address the problems in the code.

That would be nice. However

1. This PR strictly increases the warning coverage of containers. The files in which this PR disables warnings do not currently have those warnings enabled either.

2. The proponents of the new warning [are the GHC committee](https://github.com/ghc-proposals/ghc-proposals/pull/71#issuecomment-364715424). I don't think it's unreasonable to ask boot libraries to make small changes to accommodate accepted GHC proposals.

Reply:
> 1. Explicitly enable the warnings in containers's cabal file, find all code that triggers them and rewrite it
> 2. Explicitly enable the warnings in container's cabal file, find all code that triggers them and disable the warnings on a per-module basis
> 3. Explicitly _dis_able the warnings in container's cabal file and do nothing to existing code
> 
> This PR implements option 2 because it has a nice combination of easy to implement and providing actual value to containers. If you prefer 3 then please let me know. I can do that. 1 is probably more controversial since it would require actually changing some of the code of containers.

My preference would of course be option 1, because that would also demonstrate how code that currently triggers the warnings can be migrated and improved. That would also have brought https://github.com/haskell/containers/pull/591 to a proper conclusion.

But I understand that you're not trying to sell the warnings, since the decision to add them to `-Wall` has already been made.

I realize that `containers` won't be worse off after merging this PR, but the disabled warnings kind of feel like another "TODO". ;)